### PR TITLE
Add padding:1px to `td` and `th` outside tables, per spec

### PR DESCRIPTION
**Pull request by Ahmad-S792**
<pre>
Add padding:1px to `td` and `th` outside tables, per spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=255844">https://bugs.webkit.org/show_bug.cgi?id=255844</a>
rdar://problem/108716695

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web Specification [1].

[1] <a href="https://html.spec.whatwg.org/#tables-2">https://html.spec.whatwg.org/#tables-2</a>

Merge: <a href="https://chromium.googlesource.com/chromium/src/+/c5ecd6d100aeefb1b77c310beb11e458fc2d139c">https://chromium.googlesource.com/chromium/src/+/c5ecd6d100aeefb1b77c310beb11e458fc2d139c</a>

As per UA Stylesheet from web-spec, the 'td' and 'th' outside of 'table' element 
should have 1 px padding.

* Source/WebCore/css/html.css: As above
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/html-to-css-mapping-1-expected.txt: Add Test Case Expectation
* LayoutTests/fast/table/table-cell-split-expected.txt: Rebaselined
* LayoutTests/fast/table/table-row-split2-expected.txt: Rebaselined
* LayoutTests/fast/table/table-section-split2-expected.txt: Rebaselined
* LayoutTests/imported/blink/fast/css/first-letter-render-quote-expected.html: Rebaselined
* LayoutTests/platform/glib/fast/dynamic/insert-before-table-part-in-continuation-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/dynamic/insert-before-table-part-in-continuation-expected.txt: Rebaselined
* LayoutTests/fast/block/nested-renderers.html: Rebaselined
* LayoutTests/platform/ios/fast/dynamic/insert-before-table-part-in-continuation-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/dynamic/insert-before-table-part-in-continuation-expected.png: Rebaselined
</pre>

https://github.com/WebKit/WebKit/commit/144a4853ddc1523052f751fa6e2b3e5103019b7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52823 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32160 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56102 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/3546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3271 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/3546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54921 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45600 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/23993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2917 "Found 3 new test failures: fast/dynamic/insert-before-table-part-in-continuation.html, fast/table/table-row-split2.html, fast/table/table-section-split2.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1705 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3068 "Found 3 new test failures: fast/dynamic/insert-before-table-part-in-continuation.html, fast/table/table-row-split2.html, fast/table/table-section-split2.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57695 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27964 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3052 "Found 3 new test failures: fast/dynamic/insert-before-table-part-in-continuation.html, fast/table/table-row-split2.html, fast/table/table-section-split2.html (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45792 "Failed to checkout and rebase branch from PR 21270") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28938 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/html.css [hunk 1]
```diff
@@ -319,6 +319,13 @@ tr {
     border-color: inherit;
 }
 
+/* https://html.spec.whatwg.org/#tables-2 */
+/* When the td/th are inside a table (the normal case), the padding is taken
+   care of by HTMLTableCellElement::additionalPresentationAttributeStyle(). */
```
Comment (nt1m) [inline Source/WebCore/css/html.css]: This function does not exist. Do you know where this actually happens?

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: https://searchfox.org/wubkat/source/Source/WebCore/html/HTMLTableCellElement.cpp#139

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: `HTMLTableCellElement::additionalPresentationalHintStyle()`

Comment (nt1m) [reply Source/WebCore/css/html.css]: There's a slight difference in name in your comment fwiw `Attribute` vs `Hint`

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: Yes - I will update. I was hungry when I did this PR so I messed-up. :-)

Comment (nt1m) [reply Source/WebCore/css/html.css]: That being said, the Chromium approach feels like cheating for the sake of the WPT. Would be good to understand why WebKit only adds conditionally 1px padding when in a table. Is this something the spec says? If not, the rule should probably be exactly the rule in the spec: `td, th { padding: 1px; }` and we should remove these lines: https://searchfox.org/wubkat/rev/02b2f8b7a253140d82f59e38a68257431139d2e7/Source/WebCore/html/HTMLTableElement.cpp#519,524

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: Web-Spec: _"If the attribute is present but parsing the attribute's value using the [rules for parsing non-negative integers](https://html.spec.whatwg.org/#rules-for-parsing-non-negative-integers) generates an error, a default value of 1px is expected to be used for that property instead."_

Comment (nt1m) [reply Source/WebCore/css/html.css]: That's for `<table>` itself not `<td>` or `<th>`, unless I'm looking at the wrong line.

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: Looking into CSS Table Layout spec: https://drafts.csswg.org/css-tables-3/#fixup & https://drafts.csswg.org/css-tables-3/#mapping

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: This comment from Chrome bug gives more context: https://bugs.chromium.org/p/chromium/issues/detail?id=1406929#c6

Comment (Ahmad-S792) [reply Source/WebCore/css/html.css]: So in summary, what I understand overall: for normal syntax with `td` etc. in `table` element, we have C++ code to add 'cellpadding' of '1px' and handle accordingly as per web-spec but when we don't follow syntax and outside of `table`, Blink modified UA spec rule to accommodate and manage it in UA stylesheet.

**Review by nt1m — APPROVED** [review]
r=me if all tests pass

File: Source/WebCore/html/HTMLTableElement.cpp [hunk 2]
```diff
@@ -427,10 +427,8 @@ void HTMLTableElement::attributeChanged(const QualifiedName& name, const AtomStr
             m_rulesAttr = AllRules;
         break;
     case AttributeNames::cellpaddingAttr:
-        if (!newValue.isEmpty())
-            m_padding = std::max(0, parseHTMLInteger(newValue).value_or(0));
-        else
-            m_padding = 1;
+        if (auto cellPadding = parseHTMLInteger(newValue))
+            m_padding = cellpadding.value();
```
Comment (nt1m) [inline Source/WebCore/html/HTMLTableElement.cpp]:
```suggestion
            m_padding = cellPadding.value();
```

in order to build

Comment (Ahmad-S792) [reply Source/WebCore/html/HTMLTableElement.cpp]: Ops! Pushed updated commit.

**Comment by Ahmad-S792** [comment]
@nt1m - It seems that it is full of mayhem and leads to lot of test failures. Should we temporarily go back old approach similar to Blink?

File: Source/WebCore/html/HTMLTableElement.cpp [hunk 3]
```diff
@@ -427,10 +427,8 @@ void HTMLTableElement::attributeChanged(const QualifiedName& name, const AtomStr
             m_rulesAttr = AllRules;
         break;
     case AttributeNames::cellpaddingAttr:
-        if (!newValue.isEmpty())
-            m_padding = std::max(0, parseHTMLInteger(newValue).value_or(0));
-        else
-            m_padding = 1;
+        if (auto cellPadding = parseHTMLInteger(newValue))
```
Comment (nt1m) [inline Source/WebCore/html/HTMLTableElement.cpp]:
```suggestion
        if (auto cellPadding = parseHTMLNonNegativeInteger(newValue))
```

This is probably not going to fix the tests, but I noticed this is more accurate according to the spec.

Comment (Ahmad-S792) [reply Source/WebCore/html/HTMLTableElement.cpp]: Done.

**Review by nt1m — CHANGES_REQUESTED** [review]
I would prefer not landing the chromium approach

**Comment by Ahmad-S792** [comment]
Will park it for time being. :-)

**Pull request closed without merging**